src/protocol/prf_sharding/mod.rs

```diff
@@ -252,6 +273,96 @@ where
     rows_chunked_by_user
 }
 
+pub async fn attribution_and_capping_and_aggregation<C, BK, TV, F, S, SB>(
+    sh_ctx: C,
+    input_rows: Vec<PrfShardedIpaInputRow<BK, TV>>,
+    num_saturating_sum_bits: usize,
+) -> Result<Vec<Replicated<F>>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + Serializable + BasicProtocols<C, F>,
+    C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    F: PrimeField + ExtendableField,
+    TV: GaloisField,
+    BK: GaloisField,
+    //for<'u> UpgradeContext<'u, C::UpgradedContext<F>, F, RecordId>: UpgradeToMalicious<'u, BitConversionTriple<Replicated<F>>, BitConversionTriple<S>>
+{
+    // Call attribution_and_capping
+    let user_level_attributions: Vec<CappedAttributionOutputs> =
+        attribution_and_capping(sh_ctx.clone(), input_rows, num_saturating_sum_bits).await?;
+
+    let prime_field_validator = sh_ctx.narrow(&Step::BinaryValidator).validator::<F>();
+    let prime_field_m_ctx = prime_field_validator.context();
+
+    // call aggregation protocol which checks onehot and multiplies value to each bucket
+
+    do_aggregation::<_, BK, TV, F, S>(prime_field_m_ctx, user_level_attributions).await
```

Comment:
the trait bound `C: UpgradedContext<F>` is not satisfied
the trait `UpgradedContext<F>` is not implemented for `C`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B6%5D?6#file%3A%2F%2F%2FUsers%2Frichaj%2FProposals%2FIPA%2Fraw-ipa%2Fsrc%2Fprotocol%2Fprf_sharding%2Fmod.rs)
mod.rs(300, 5): required by a bound introduced by this call
mod.rs(308, 8): required by a bound in `do_aggregation`
mod.rs(282, 25): consider further restricting this bound: ` + protocol::context::UpgradedContext<F>`